### DISCUSSION TITLE
how to customze kong gateway logs  (helm install method)

### DISCUSSION THREAD

**Discussion by linzj3** [Kong/kong#13750] [Help] [answered]
i use helm to install kong, i find that i cannot start kong-dp normal  if i try to customze kong gateway logs.
kubectl get cm -n kong  custom-format-file -oyaml
apiVersion: v1
data:
  custom_format.conf: |
    log_format custom_format 'service_name|1|1|$http_x_forwarded_for|$server_addr|$server_port|$http_traceId|$msec|$request_time|$request_method|$request_uri|$status|$request_length|$body_bytes_sent|$http_referer|$http_user_agent|$upstream_addr|$http_host';
kind: ConfigMap
metadata:
  creationTimestamp: "2024-10-14T08:41:45Z"
  name: custom-format-file
  namespace: kong
  resourceVersion: "40137522"
  uid: 4c86e423-030a-4778-8064-94453c7fee87
this is my values.yaml for kong gateway:
[sit-eu_values-dp.zip](https://github.com/user-attachments/files/17362440/sit-eu_values-dp.zip)
this is pod logs: 
[kong-kong-dp-kong-794857dbf7-tkg2m-1728891937666527289.log](https://github.com/user-attachments/files/17362449/kong-kong-dp-kong-794857dbf7-tkg2m-1728891937666527289.log)

now the dataplane of kong-gateway cannot run normal. how i can do to solve this problem?  hope someone help me, thank you

**Comment by linzj3** [comment]
kong@kong-dp-kong-74c888c65f-vzk2n:/kong_prefix$ more nginx.conf 
pid pids/nginx.pid;

error_log /dev/stderr notice;

# injected nginx_main_* directives
daemon off;
user kong kong;
worker_processes 2;
worker_rlimit_nofile 16384;

include 'nginx-inject.conf';

events {
    # injected nginx_events_* directives
    multi_accept on;
    worker_connections 16384;
}


http {
    include 'nginx-kong.conf';
}

kong@kong-dp-kong-74c888c65f-vzk2n:/kong_prefix$ cat nginx-kong.conf 
server_tokens off;

error_log /dev/stderr notice;

lua_package_path       '/opt/?.lua;/opt/?/init.lua;;;;;';
lua_package_cpath      ';;;';
lua_socket_pool_size   256;
lua_socket_log_errors  off;
lua_max_running_timers 4096;
lua_max_pending_timers 16384;

include 'nginx-kong-inject.conf';

lua_shared_dict kong                        5m;
lua_shared_dict kong_locks                  8m;
lua_shared_dict kong_healthchecks           5m;
lua_shared_dict kong_cluster_events         5m;
lua_shared_dict kong_rate_limiting_counters 12m;
lua_shared_dict kong_core_db_cache          128m;
lua_shared_dict kong_core_db_cache_miss     12m;
lua_shared_dict kong_db_cache               128m;
lua_shared_dict kong_db_cache_miss          12m;
lua_shared_dict kong_secrets                5m;


underscores_in_headers on;
ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;

# injected nginx_http_* directives
charset UTF-8;
client_body_buffer_size 8k;
client_max_body_size 0;
include /log_format/custom_format.conf;
keepalive_requests 10000;
lua_regex_cache_max_entries 8192;
lua_regex_match_limit 100000;
lua_shared_dict prometheus_metrics 5m;
ssl_dhparam /kong_prefix/ssl/ffdhe2048.pem;
ssl_prefer_server_ciphers off;
ssl_protocols TLSv1.2 TLSv1.3;
ssl_session_tickets on;
ssl_session_timeout 1d;

uninitialized_variable_warn  off;

init_by_lua_block {
    Kong = require 'kong'
    Kong.init()
}

init_worker_by_lua_block {
    Kong.init_worker()
}

exit_worker_by_lua_block {
    Kong.exit_worker()
}

log_format kong_log_format '$remote_addr - $remote_user [$time_local] '
                           '"$request" $status $body_bytes_sent '
                           '"$http_referer" "$http_user_agent" '
                           'kong_request_id: "$kong_request_id"';

# Load variable indexes
lua_kong_load_var_index default;

upstream kong_upstream {
    server 0.0.0.1;

    # injected nginx_upstream_* directives

    balancer_by_lua_block {
        Kong.balancer()
    }
}

server {
    server_name kong;
    listen 0.0.0.0:8000;
    listen [0000:0000:0000:0000:0000:0000:0000:0000]:8000;
    listen 0.0.0.0:8443 ssl;
    listen [0000:0000:0000:0000:0000:0000:0000:0000]:8443 ssl;

    http2 on;

    error_page 400 404 405 408 411 412 413 414 417 /kong_error_handler;
    error_page 494 =494                            /kong_error_handler;
    error_page 500 502 503 504                     /kong_error_handler;

    # Append the kong request id to the error log
    # https://github.com/Kong/lua-kong-nginx-module#lua_kong_error_log_request_id
    lua_kong_error_log_request_id $kong_request_id;

    access_log /dev/stdout kong_log_format;

    error_log  /dev/stderr notice;

    ssl_certificate     /kong_prefix/ssl/kong-default.crt;
    ssl_certificate_key /kong_prefix/ssl/kong-default.key;
    ssl_certificate     /kong_prefix/ssl/kong-default-ecdsa.crt;
    ssl_certificate_key /kong_prefix/ssl/kong-default-ecdsa.key;
    ssl_session_cache   shared:SSL:10m;
    ssl_certificate_by_lua_block {
        Kong.ssl_certificate()
    }

    # injected nginx_proxy_* directives
    real_ip_header x-forwarded-for;
    real_ip_recursive on;
    set_real_ip_from 0.0.0.0/0;

    rewrite_by_lua_block {
        Kong.rewrite()
    }

    access_by_lua_block {
        Kong.access()
    }

    header_filter_by_lua_block {
        Kong.header_filter()
    }

    body_filter_by_lua_block {
        Kong.body_filter()
    }

    log_by_lua_block {
        Kong.log()
    }

    location / {
        default_type                     '';

        set $ctx_ref                     '';
        set $upstream_te                 '';
        set $upstream_via                '';
        set $upstream_host               '';
        set $upstream_upgrade            '';
        set $upstream_connection         '';
        set $upstream_scheme             '';
        set $upstream_uri                '';
        set $upstream_x_forwarded_for    '';
        set $upstream_x_forwarded_proto  '';
        set $upstream_x_forwarded_host   '';
        set $upstream_x_forwarded_port   '';
        set $upstream_x_forwarded_path   '';
        set $upstream_x_forwarded_prefix '';
        set $kong_proxy_mode             'http';

        proxy_http_version      1.1;
        proxy_buffering          on;
        proxy_request_buffering  on;

        # injected nginx_location_* directives

        proxy_set_header      TE                 $upstream_te;
        proxy_set_header      Via                $upstream_via;
        proxy_set_header      Host               $upstream_host;
        proxy_set_header      Upgrade            $upstream_upgrade;
        proxy_set_header      Connection         $upstream_connection;
        proxy_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        proxy_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        proxy_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        proxy_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        proxy_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        proxy_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        proxy_set_header      X-Real-IP          $remote_addr;
        proxy_set_header      X-Kong-Request-Id  $kong_request_id;
        proxy_pass_header     Server;
        proxy_pass_header     Date;
        proxy_ssl_name        $upstream_host;
        proxy_ssl_server_name on;
        proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
    }

    location @unbuffered {
        internal;
        default_type         '';
        set $kong_proxy_mode 'unbuffered';

        proxy_http_version      1.1;
        proxy_buffering         off;
        proxy_request_buffering off;

        proxy_set_header      TE                 $upstream_te;
        proxy_set_header      Via                $upstream_via;
        proxy_set_header      Host               $upstream_host;
        proxy_set_header      Upgrade            $upstream_upgrade;
        proxy_set_header      Connection         $upstream_connection;
        proxy_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        proxy_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        proxy_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        proxy_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        proxy_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        proxy_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        proxy_set_header      X-Real-IP          $remote_addr;
        proxy_set_header      X-Kong-Request-Id  $kong_request_id;
        proxy_pass_header     Server;
        proxy_pass_header     Date;
        proxy_ssl_name        $upstream_host;
        proxy_ssl_server_name on;
        proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
    }

    location @unbuffered_request {
        internal;
        default_type         '';
        set $kong_proxy_mode 'unbuffered';

        proxy_http_version      1.1;
        proxy_buffering          on;
        proxy_request_buffering off;

        proxy_set_header      TE                 $upstream_te;
        proxy_set_header      Via                $upstream_via;
        proxy_set_header      Host               $upstream_host;
        proxy_set_header      Upgrade            $upstream_upgrade;
        proxy_set_header      Connection         $upstream_connection;
        proxy_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        proxy_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        proxy_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        proxy_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        proxy_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        proxy_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        proxy_set_header      X-Real-IP          $remote_addr;
        proxy_set_header      X-Kong-Request-Id  $kong_request_id;
        proxy_pass_header     Server;
        proxy_pass_header     Date;
        proxy_ssl_name        $upstream_host;
        proxy_ssl_server_name on;
        proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
    }

    location @unbuffered_response {
        internal;
        default_type         '';
        set $kong_proxy_mode 'unbuffered';

        proxy_http_version      1.1;
        proxy_buffering         off;
        proxy_request_buffering  on;

        proxy_set_header      TE                 $upstream_te;
        proxy_set_header      Via                $upstream_via;
        proxy_set_header      Host               $upstream_host;
        proxy_set_header      Upgrade            $upstream_upgrade;
        proxy_set_header      Connection         $upstream_connection;
        proxy_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        proxy_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        proxy_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        proxy_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        proxy_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        proxy_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        proxy_set_header      X-Real-IP          $remote_addr;
        proxy_set_header      X-Kong-Request-Id  $kong_request_id;
        proxy_pass_header     Server;
        proxy_pass_header     Date;
        proxy_ssl_name        $upstream_host;
        proxy_ssl_server_name on;
        proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
    }

    location @grpc {
        internal;
        default_type         '';
        set $kong_proxy_mode 'grpc';

        grpc_set_header      TE                 $upstream_te;
        grpc_set_header      Via                $upstream_via;
        grpc_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        grpc_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        grpc_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        grpc_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        grpc_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        grpc_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        grpc_set_header      X-Real-IP          $remote_addr;
        grpc_set_header      X-Kong-Request-Id  $kong_request_id;
        grpc_pass_header     Server;
        grpc_pass_header     Date;
        grpc_ssl_name        $upstream_host;
        grpc_ssl_server_name on;
        grpc_pass            $upstream_scheme://kong_upstream;
    }

    location = /kong_buffered_http {
        internal;
        default_type         '';
        set $kong_proxy_mode 'http';

        rewrite_by_lua_block       {
          -- ngx.location.capture will create a new nginx request,
          -- so the upstream ssl-related info attached to the `r` gets lost.
          -- we need to re-set them here to the new nginx request.
          local ctx = ngx.ctx
          local upstream_ssl = require("kong.runloop.upstream_ssl")

          upstream_ssl.set_service_ssl(ctx)
          upstream_ssl.fallback_upstream_client_cert(ctx)
        }
        access_by_lua_block        {;}
        header_filter_by_lua_block {;}
        body_filter_by_lua_block   {;}
        log_by_lua_block           {;}

        proxy_http_version 1.1;
        proxy_set_header      TE                 $upstream_te;
        proxy_set_header      Via                $upstream_via;
        proxy_set_header      Host               $upstream_host;
        proxy_set_header      Upgrade            $upstream_upgrade;
        proxy_set_header      Connection         $upstream_connection;
        proxy_set_header      X-Forwarded-For    $upstream_x_forwarded_for;
        proxy_set_header      X-Forwarded-Proto  $upstream_x_forwarded_proto;
        proxy_set_header      X-Forwarded-Host   $upstream_x_forwarded_host;
        proxy_set_header      X-Forwarded-Port   $upstream_x_forwarded_port;
        proxy_set_header      X-Forwarded-Path   $upstream_x_forwarded_path;
        proxy_set_header      X-Forwarded-Prefix $upstream_x_forwarded_prefix;
        proxy_set_header      X-Real-IP          $remote_addr;
        proxy_set_header      X-Kong-Request-Id  $kong_request_id;
        proxy_pass_header     Server;
        proxy_pass_header     Date;
        proxy_ssl_name        $upstream_host;
        proxy_ssl_server_name on;
        proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
    }

    location = /kong_error_handler {
        internal;

        default_type                 '';

        rewrite_by_lua_block {;}
        access_by_lua_block  {;}

        content_by_lua_block {
            Kong.handle_error()
        }
    }
}

server {
    charset UTF-8;
    server_name kong_admin;
    listen 127.0.0.1:8444 ssl;
    listen [0000:0000:0000:0000:0000:0000:0000:0001]:8444 ssl;

    http2 on;

    access_log /dev/stdout;
    error_log  /dev/stderr notice;

    ssl_certificate     /kong_prefix/ssl/admin-kong-default.crt;
    ssl_certificate_key /kong_prefix/ssl/admin-kong-default.key;
    ssl_certificate     /kong_prefix/ssl/admin-kong-default-ecdsa.crt;
    ssl_certificate_key /kong_prefix/ssl/admin-kong-default-ecdsa.key;
    ssl_session_cache   shared:AdminSSL:10m;

    # injected nginx_admin_* directives
    client_body_buffer_size 10m;
    client_max_body_size 10m;

    location / {
        default_type application/json;
        content_by_lua_block {
            Kong.admin_content()
        }
        header_filter_by_lua_block {
            Kong.admin_header_filter()
        }
    }

    location /robots.txt {
        return 200 'User-agent: *\nDisallow: /';
    }
}

server {
    charset UTF-8;
    server_name kong_status;
    listen 0.0.0.0:8100;
    listen [0000:0000:0000:0000:0000:0000:0000:0000]:8100;


    access_log off;
    error_log  /dev/stderr notice;


    # injected nginx_status_* directives

    location / {
        default_type application/json;
        content_by_lua_block {
            Kong.status_content()
        }
        header_filter_by_lua_block {
            Kong.status_header_filter()
        }
    }

    location /robots.txt {
        return 200 'User-agent: *\nDisallow: /';
    }
}

server {
    server_name kong_gui;
    listen 0.0.0.0:8002;
    listen [0000:0000:0000:0000:0000:0000:0000:0000]:8002;
    listen 0.0.0.0:8445 ssl;
    listen [0000:0000:0000:0000:0000:0000:0000:0000]:8445 ssl;

    http2 on;

    ssl_certificate     /kong_prefix/ssl/admin-gui-kong-default.crt;
    ssl_certificate_key /kong_prefix/ssl/admin-gui-kong-default.key;
    ssl_certificate     /kong_prefix/ssl/admin-gui-kong-default-ecdsa.crt;
    ssl_certificate_key /kong_prefix/ssl/admin-gui-kong-default-ecdsa.key;
    ssl_protocols TLSv1.2 TLSv1.3;

    client_max_body_size 10m;
    client_body_buffer_size 10m;

    types {
        text/html                             html htm shtml;
        text/css                              css;
        text/xml                              xml;
        image/gif                             gif;
        image/jpeg                            jpeg jpg;
        application/javascript                js;
        application/json                      json;
        image/png                             png;
        image/tiff                            tif tiff;
        image/x-icon                          ico;
        image/x-jng                           jng;
        image/x-ms-bmp                        bmp;
        image/svg+xml                         svg svgz;
        image/webp                            webp;
    }

    access_log /dev/stdout;
    error_log /dev/stderr;

    gzip on;
    gzip_types text/plain text/css application/json application/javascript;

    include nginx-kong-gui-include.conf;
}


server {
    charset UTF-8;
    server_name kong_worker_events;
    listen unix:/kong_prefix/sockets/we;
    access_log off;
    location / {
        content_by_lua_block {
          require("resty.events.compat").run()
        }
    }
}

**Comment by ProBrian** [comment]
@randmonkey Could you help to take a look of this issue?

**Comment by linzj3** [comment, marked as answer]
i take two of env in values.yaml, it's mistake....  change to like it:
env:
  # data_plane nodes do not have a database
  role: data_plane
  database: "off"
  log_level: notice
  router_flavor: "traditional_compatible"
  nginx_worker_processes: "4"
  proxy_access_log: /dev/stdout
  proxy_access_log: /dev/stdout custom_format
  #nginx_http_log_format: custom_format escape=json '{"@timestamp":"$time_iso8601","host":"$host","request":"$request","request_body":"$request_body","remote_addr":"$remote_addr","uri":"$uri","upstream_addr":"$upstream_addr","request_time":"$request_time","upstream_status":"$upstream_status","upstream_header_time":"$upstream_header_time","upstream_response_time":"$upstream_response_time","status":"$status"}'
  nginx_http_include: /tmp/custom_format.conf
  admin_access_log: /dev/stdout
  admin_gui_access_log: /dev/stdout
  portal_api_access_log: /dev/stdout
  proxy_error_log: /dev/stderr
  admin_error_log: /dev/stderr
  admin_gui_error_log: /dev/stderr
  portal_api_error_log: /dev/stderr
  real_ip_header: x-forwarded-for
  real_ip_recursive: "on"
  trusted_ips: "0.0.0.0/0"

it can do well
